### PR TITLE
Refactor/160 normalize questionnaire

### LINKED ISSUE #160: Apply data store refactor to questionnaire editor
See #154

### PR DESCRIPTION
**Pull request by nekron**
Fixes #160

### CODE REVIEW

File: web/static/js/actions/questionnaire.js [hunk 1]
```diff
@@ -115,12 +115,11 @@ export const addStep = (stepType) => {
 
 export const save = () => {
   return (dispatch, getState) => {
-    const questionnaire = getState().questionnaire
-
-    if (questionnaire.id == null) {
-      api.createQuestionnaire(questionnaire.projectId, questionnaire)
-    } else {
+    const questionnaire = getState().questionnaire.data
+    if (questionnaire.id) {
```
Comment (mverzilli) [inline web/static/js/actions/questionnaire.js]: Why this instead of `questionnaire.id == null`? I'm a bit weary of falsiness in JS

File: test/js/reducers/questionnaire.spec.js [hunk 2]
```diff
@@ -285,7 +285,7 @@ describe('questionnaire reducer', () => {
   })
 })
 
-const questionnaire = {
+const questionnaire = () => ({
```
Comment (mverzilli) [inline test/js/reducers/questionnaire.spec.js]: (Out of curiosity) Why this? The only reason I can think of is immutability, is that it?

File: web/static/js/components/questionnaires/QuestionnaireEditor.jsx [hunk 3]
```diff
@@ -77,8 +71,8 @@ class QuestionnaireEditor extends Component {
     // I couldn't find a better way. Ideally this should be a sort of "callback"
     // to the addStep method, without involving additional component state handling
     // or explicit management via Redux reducers.
-    const questionnaireData = newProps.questionnaire.data
-    if (this.state.addingStep && questionnaireData != null && questionnaireData.steps != null && questionnaireData.steps.length > 0) {
+    const questionnaireData = newProps.questionnaire
+    if (this.state.addingStep && questionnaireData && questionnaireData.steps != null && questionnaireData.steps.length > 0) {
```
Comment (mverzilli) [inline web/static/js/components/questionnaires/QuestionnaireEditor.jsx]: Ditto falsiness

File: test/js/reducers/questionnaire.spec.js (hunk 2 above)
Comment (nekron) [inline test/js/reducers/questionnaire.spec.js]: Yup. Eventually we solved that by using deepFreeze.

File: web/static/js/components/questionnaires/QuestionnaireEditor.jsx (hunk 3 above)
Comment (nekron) [inline web/static/js/components/questionnaires/QuestionnaireEditor.jsx]: I'll make a regression over this once it's merged and see if there is a breaking case.

File: web/static/js/actions/questionnaire.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+export const ADD_STEP = 'QUESTIONNAIRE_ADD_STEP'
+export const DELETE_STEP = 'QUESTIONNAIRE_DELETE_STEP'
+export const CHANGE_STEP_TITLE = 'QUESTIONNAIRE_CHANGE_STEP_TITLE'
+export const CHANGE_STEP_PROMPT_SMS = 'QUESTIONNAIRE_CHANGE_STEP_PROMPT_SMS'
+export const CHANGE_STEP_STORE = 'QUESTIONNAIRE_CHANGE_STEP_STORE'
+export const ADD_CHOICE = 'QUESTIONNAIRE_EDITOR_ADD_CHOICE'
+export const DELETE_CHOICE = 'QUESTIONNAIRE_EDITOR_DELETE_CHOICE'
+export const CHANGE_CHOICE = 'QUESTIONNAIRE_EDITOR_CHANGE_CHOICE'
+
+export const fetchQuestionnaire = (projectId, questionnaireId) => (dispatch, getState) => {
```
Comment (asterite) [inline web/static/js/actions/questionnaire.js]: I don't understand why there is this and `if needed` variants. The logic is covered by the `fetching` and `filter` properties, and #154 doesn't mention two ways of fetching a questionnaire. Why is this needed?

File: web/static/js/components/questionnaires/QuestionnaireEditor.jsx [hunk 5]
```diff
@@ -39,7 +39,45 @@ class QuestionnaireEditor extends Component {
...
+
+  internalState(currentStep, addingStep = false) {
+    return {
+      currentStep,
+      addingStep
+    }
+  }
+
+  componentWillReceiveProps(newProps) {
+    // This feels a bit hacky, but it let's us expand the step we just created.
```
Comment (asterite) [inline web/static/js/components/questionnaires/QuestionnaireEditor.jsx]: If the stat stores what step is expanded (like it was before) it's super simple to implement and there's no need for a hack. Why not implement it that way?

File: web/static/js/actions/questionnaire.js (hunk 4 above)
Comment (nekron) [inline web/static/js/actions/questionnaire.js]: There is no need to export this function. It's just to be able to force it to fetch if necessary.

File: web/static/js/components/surveys/SurveyEdit.jsx [hunk 6]
```diff
@@ -29,9 +29,8 @@ class SurveyEdit extends Component {
     const { dispatch, projectId, surveyId, router } = this.props
     if (projectId && surveyId) {
       projectActions.fetchProject(projectId)
-      dispatch(actions.fetch(projectId, surveyId))
+      dispatch(actions.fetchSurveyIfNeeded(projectId, surveyId))
```
Comment (nekron) [inline web/static/js/components/surveys/SurveyEdit.jsx]: The first is just there to simplify the reading of the second.

It's being exported just in case that someone need to force the fetch for some reason. Maybe we can keep it private for now until there is a good reason for that.

File: docker-compose.yml [hunk 7]
```diff
@@ -34,6 +34,7 @@ services:
   brunch:
     image: node:6.3.1
     command: node_modules/brunch/bin/brunch watch
+    restart: always
```
Comment (nekron) [inline docker-compose.yml]: True.

It was useful as we developed, and there were already some conflicting changes in master that would've led to a hard to test scenario. That prevented us from making this change in master and then merging it.

I should've kept it stashed.

File: web/static/js/actions/questionnaire.js (hunk 4 above)
Comment (asterite) [inline web/static/js/actions/questionnaire.js]: Ah, I see. So maybe the public API should be `fetchQuestionnaire` and the private one be `forceFetchQuestionnaire`?

File: web/static/js/components/questionnaires/QuestionnaireEditor.jsx (hunk 5 above)
Comment (nekron) [inline web/static/js/components/questionnaires/QuestionnaireEditor.jsx]: We should talk this with @mverzilli

Comment (mverzilli) [inline web/static/js/components/questionnaires/QuestionnaireEditor.jsx]: It felt wrong because we'd have to add ad-hoc questionnaire substate to the store for the editor. This way we keep the reducer pretty focused on data management, and the editor manages UI state. 

But maybe it's not worth the effort, since as you said the previous approach seemed simpler. 

The questionnaire editor will suffer a lot of changes in the short term, so maybe we can move this to it's own issue.
